data/transactions/logic/assembler.go

```diff
@@ -1086,79 +1111,176 @@ func asmDefault(ops *OpStream, spec *OpSpec, args []string) error {
 				}
 				ops.pending.WriteByte(val)
 			}
+		case immInt8:
+			val, err := int8Imm(args[i], imm.Name)
+			if err != nil {
+				return ops.errorf("%s %w", spec.Name, err)
+			}
+			ops.pending.WriteByte(val)
 		default:
 			return ops.errorf("unable to assemble immKind %d", imm.kind)
 		}
 	}
 	return nil
 }
 
-// Interprets the arg at index argIndex as byte-long immediate
-func getByteImm(args []string, argIndex int) (byte, bool) {
+// getImm interprets the arg at index argIndex as an immediate
+func getImm(args []string, argIndex int) (int, bool) {
 	if len(args) <= argIndex {
 		return 0, false
 	}
-	n, err := strconv.ParseUint(args[argIndex], 0, 8)
+	// We want to parse anything from -128 up to 255. So allow 9 bits.
+	// Normal assembly checkign willl catch signed as byte, vice versa
+	n, err := strconv.ParseInt(args[argIndex], 0, 9)
 	if err != nil {
 		return 0, false
 	}
-	return byte(n), true
+	return int(n), true
 }
 
-func typeSwap(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes) {
-	topTwo := StackTypes{StackAny, StackAny}
+func anyTypes(n int) StackTypes {
+	as := make(StackTypes, n)
+	for i := range as {
+		as[i] = StackAny
+	}
+	return as
+}
+
+func typeSwap(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes, error) {
+	swapped := StackTypes{StackAny, StackAny}
 	top := len(pgm.stack) - 1
 	if top >= 0 {
-		topTwo[1] = pgm.stack[top]
+		swapped[0] = pgm.stack[top]
 		if top >= 1 {
-			topTwo[0] = pgm.stack[top-1]
+			swapped[1] = pgm.stack[top-1]
 		}
 	}
-	reversed := StackTypes{topTwo[1], topTwo[0]}
-	return nil, reversed
+	return nil, swapped, nil
 }
 
-func typeDig(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes) {
-	n, ok := getByteImm(args, 0)
+func typeDig(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes, error) {
+	n, ok := getImm(args, 0)
 	if !ok {
-		return nil, nil
-	}
-	depth := int(n) + 1
-	anys := make(StackTypes, depth)
-	returns := make(StackTypes, depth+1)
-	for i := range anys {
-		anys[i] = StackAny
-		returns[i] = StackAny
+		return nil, nil, nil
```

Comment:
nit: if it is not ok returned from `getImm`, then I don't think you should return error as `nil`, right?
For `getImm` return `ok = false` when immediate arg parse not successful.

Reply:
the `typed*` functions are a kind of "precheck" that only some opcodes perform.  All the opcodes have their normal handling performed later. So, rather than reproduce that error message here, we just say that this function didn't catch any extra problems, and let the normal checking take care of it.

Reply:
ok, then I think it just makes sense as a precheck, and I assume the `error` in return is for the sake of `refineFunc` type definition in `stacky` and other related funcs?

Reply:
Yes, the idea is that if the `refineFunc` finds an error that normal opcode checking would not detect, it can return such an error directly.